src/solvers/jesse_dg/ESDG_2D.jl

```diff
@@ -0,0 +1,281 @@
+QshskewTr = Matrix(Qshskew')
+
+# StructArray initialization - problem in entropy2cons
+staticzip(::Type{SVector{M}},x::NTuple{M,Array{T,N}}) where {M,T,N} = SVector{M}.(zip(x...))
+# xyz = StructArray(staticzip(SVector{2},md.xyz))
+# U = (xyz->initial_condition(xyz,0,eqn)).(xyz) # hack to retain tuple elements in U
+
+## ====== workaround for StructArrays/DiffEq.jl from https://github.com/SciML/OrdinaryDiffEq.jl/issues/1386
+function RecursiveArrayTools.recursivecopy(a::AbstractArray{T,N}) where {T<:AbstractArray,N}
+    if ArrayInterface.ismutable(a)
+      b = similar(a)
+      map!(recursivecopy,b,a)
+    else
+      ArrayInterface.restructure(a,map(recursivecopy,a))
+    end
+end
+ArrayInterface.ismutable(x::StructArray) = true
+## ====== end workaround ===========
+
+Base.real(rd::RefElemData) = Float64 # is this for DiffEq.jl?
+
+function Trixi.create_cache(mesh::UnstructuredMesh, equations, rd::RefElemData, RealT, uEltype)
+
+    @unpack VXYZ,EToV = mesh
+    md = MeshData(VXYZ...,EToV,rd)
+    md = make_periodic(md,rd)
+
+    # for flux differencing on general elements
+    Qrhskew,Qshskew,VhP,Ph = build_hSBP_ops(rd)
+    QrhskewTr = Matrix(Qrhskew')
+    QshskewTr = Matrix(Qshskew')
+
+    # tmp variables for entropy projection
+    Uq = StructArray(staticzip(SVector{4},ntuple(_->similar(md.xq),4)))
+    VUq = similar(Uq)
+    VUh = StructArray(staticzip(SVector{4},ntuple(_->similar([md.xq;md.xf]),4)))
+    Uh = similar(VUh)
+
+    cache = (;md,
+            QrhskewTr,QshskewTr,VhP,Ph,
+            Uq,VUq,VUh,Uh)
+
+    return cache
+end
+
+@inline function tmap!(f,out,x)
+    Trixi.@threaded for i = 1:length(x)
+        @inbounds out[i] = f(x[i])
+    end
+end
+
+project_and_store!(y,x) = mul!(y,Ph,x) # can't use matmul! b/c its applied to a subarray
+
+## workaround for matmul! with threads https://discourse.julialang.org/t/odd-benchmarktools-timings-using-threads-and-octavian/59838/5
+@inline function bmap!(f,out,x)
+    @batch for i = 1:length(x)
+        @inbounds out[i] = f(x[i])
+    end
+end
+
+@inline function compute_entropy_projection!(Q,rd::RefElemData,cache,eqn) 
+    @unpack Vq = rd    
+    @unpack VhP,Ph = cache
+    @unpack Uq, VUq, VUh, Uh = cache
+
+    # entropy projection - should be zero alloc
+    StructArrays.foreachfield((uout,u)->mul!(uout,Vq,u),Uq,Q) # matmul causes issue?
+    tmap!(u->cons2entropy(u,eqn),VUq,Uq) # 77.5μs
+    StructArrays.foreachfield((uout,u)->mul!(uout,VhP,u),VUh,VUq) # matmul slower 
+    tmap!(v->entropy2cons(v,eqn),Uh,VUh) # 327.204 μs
+
+    # StructArrays.foreachfield((uout,u)->matmul!(uout,Vq,u),Uq,Q)  
+    # bmap!(u->cons2entropy(u,eqn),VUq,Uq) # 77.5μs
+    # StructArrays.foreachfield((uout,u)->matmul!(uout,VhP,u),VUh,VUq) 
+    # bmap!(v->entropy2cons(v,eqn),Uh,VUh) # 327.204 μs
+
+    Nh,Nq = size(VhP)
+    Uf = view(Uh,Nq+1:Nh,:) # 24.3 μs
+
+    return Uh,Uf
+end
+
+@inline function max_abs_speed_normal(UL, UR, normal, equations::CompressibleEulerEquations2D) 
```

Comment:
Can you use Trixi's function but with `normal::AbstractVector`?

Reply:
Yeah - it's not the same flux, but it works fine too.